docs/ADR-8-Use-RIO-in-cardano‐cli.md

```diff
@@ -0,0 +1,181 @@
+
+![image](https://github.com/user-attachments/assets/6a50e8f2-9140-4b7e-afa5-ff778d696742)
+ 
+- The ExceptT IO is a known anti-pattern for these reasons and others as per:
+  - https://github.com/haskell-effectful/effectful/blob/master/transformers.md
+  - https://tech.fpcomplete.com/blog/2016/11/exceptions-best-practices-haskell/
+
+
+# Proposed Solution
+
+I propose to replace  `ExceptT someError IO a` with [RIO env a](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#t:RIO).
+
+
+### Example 
+
+Below is how `cardano-cli` is currently structured. `ExceptT` with errors wrapping errors. However the errors ultimately end up being rendered at the top level to the user.
+
+```haskell
+
+-- TOP LEVEL -- 
+
+data ExampleClientCommand = ClientCommandTransactions ClientCommandTransactions
+
+data ExampleClientCommandErrors
+  = CmdError CmdError
+-- | ByronClientError ByronClientCmdError
+-- | AddressCmdError AddressCmdError
+-- ...
+data CmdError
+  = ExampleTransactionCmdError ExampleTransactionCmdError
+-- | AddressCommand AnyEraCommand
+-- | ByronCommand AddressCmds
+-- ...
+
+topLevelRunCommand :: ExampleClientCommand -> ExceptT ExampleClientCommandErrors IO ()
+topLevelRunCommand (ClientCommandTransactions txsCmd) =
+  firstExceptT (CmdError . ExampleTransactionCmdError) $ runClientCommandTransactions txsCmd
+
+-- SUB LEVEL --
+
+data ClientCommandTransactions = DummyClientCommandToRun
+
+data ExampleTransactionCmdError
+  = TransactionWriteFileError !(FileError ())
+
+runClientCommandTransactions
+  :: ()
+  => ClientCommandTransactions
+  -> ExceptT ExampleTransactionCmdError IO ()
+runClientCommandTransactions DummyClientCommandToRun =
+  ...
+  left $
+    TransactionWriteFileError $
+      FileError "dummy.file" ()
+``` 
+
+Proposed change:
+
+```haskell
+data ClientCommandTransactions = DummyClientCommandToRun
+
+data ExampleClientCommand = ClientCommandTransactions ClientCommandTransactions
+
+topLevelRunCommand :: ExampleClientCommand -> RIO () ()
+topLevelRunCommand (ClientCommandTransactions txsCmd) =
+  runClientCommandTransactions txsCmd
+
+runClientCommandTransactions
+  :: HasCallStack
+  => ClientCommandTransactions
+  -> RIO () ()
+runClientCommandTransactions DummyClientCommandToRun =
+  ...
+  throwIO $
+    CustomCliException $
+      FileError "dummy.file" ()
+```
+
+We have eliminated `data ExampleClientCommandErrors` and `data CmdError` and improved the composability of our code.
+
+### Pros 
+- Additional [logging functionality](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#g:5)
+- Explicit environment dependencies e.g `logError :: HasLogFunc env => Text -> RIO env ()`
+- Better composability i.e no more errors that wrap errors (see above).
```

Comment:
```suggestion
- Better composability i.e no more errors that wrap errors (see above).
- Less bookeping required: no more wrapping low-level errors into high-level errors
```

Reply:
I think the point above already captures this